### PR TITLE
Feature: Cover products contract

### PR DESCRIPTION
**Pull request by danoctavian**
## Context

Reduce size of Cover to fit in more functionality.

Issue: https://github.com/NexusMutual/smart-contracts/issues/885

## Changes proposed in this pull request

Move functions and data structures related to product and product types to a new contract called CoverProducts.sol

move createStakingPool to StakingProducts.sol

Leaving createStakingPool where it is is not feasible because of Cover.sol contract size. Enough Head room is needed

## Test plan

New tests are not needed just existing ones moved to different files.

## Checklist

- [x] Rebased the base branch
- [x] Attached corresponding Github issue
- [x] Prefixed the name with the type of change (i.e. feat, chore, test)
- [x] Performed a self-review of my own code
- [x] Followed the style guidelines of this project
- [x] Made corresponding changes to the documentation
- [x] Didn't generate new warnings
- [x] Didn't generate failures on existing tests
- [x] Added tests that prove my fix is effective or that my feature works

## Review

When reviewing a PR, please indicate intention in comments using the following emojis:
* :cake: = Nice to have but not essential.
* :bulb: = Suggestion or a comment based on personal opinion
* :hammer: = I believe this should be changed.
* :thinking: = I don’t understand something, do you mind giving me more context?
* :rocket: = Feedback

### CODE REVIEW

File: contracts/modules/staking/StakingProducts.sol [hunk 1]
```diff
@@ -599,8 +605,75 @@ contract StakingProducts is IStakingProducts, MasterAwareV2, Multicall {
...
+
+  function coverProducts() internal view returns (ICoverProducts) {
+    return ICoverProducts(getInternalContractAddress(ID.CP));
+  }
+
   function changeDependentContractAddress() external {
     // none :)
   }
 
+  function createStakingPool(
```
Comment (danoctavian) [inline contracts/modules/staking/StakingProducts.sol:624]: i can't recall what was the latest discussion on this function

happy to move it to CoverProducts if it makes sense.

File: contracts/modules/legacy/LegacyPooledStaking.sol [hunk 2]
```diff
@@ -841,154 +841,6 @@ contract LegacyPooledStaking is IPooledStaking, MasterAwareV2, PricesV1 {
     return products;
   }
 
-  function createPoolAndDeposit(
```
Comment (danoctavian) [inline contracts/modules/legacy/LegacyPooledStaking.sol:844]: this is no longer used - it was for migration only.

As opposed to updating the code for the new structure to compile - i'd rather delete it.

File: contracts/modules/cover/CoverProducts.sol [hunk 3]
```diff
@@ -0,0 +1,220 @@
...
+import "../../interfaces/IStakingPoolBeacon.sol";
+import "../../interfaces/IStakingPoolFactory.sol";
+import "../../interfaces/ITokenController.sol";
+import "../../libraries/Math.sol";
+import "../../libraries/SafeUintCast.sol";
+import "../../libraries/StakingPoolLibrary.sol";
+import "../../interfaces/IStakingProducts.sol";
+import "../../interfaces/ICoverProducts.sol";
+
+contract CoverProducts is ICoverProducts, MasterAwareV2, Multicall {
```
Comment (danoctavian) [inline contracts/modules/cover/CoverProducts.sol:25]: TODO: need a migration function for old products to this.

File: contracts/modules/cover/Cover.sol [hunk 4]
```diff
@@ -534,54 +531,6 @@ contract Cover is ICover, MasterAwareV2, IStakingPoolBeacon, ReentrancyGuard, Mu
...
-      isPrivatePool,
-      initialPoolFee,
-      maxPoolFee,
-      poolId,
-      ipfsDescriptionHash
-    );
-
-    tokenController().assignStakingPoolManager(poolId, msg.sender);
-
-    stakingProducts().setInitialProducts(poolId, productInitParams);
```
Comment (roxdanila) [inline contracts/modules/cover/Cover.sol:580]: 💡 idea: 
1. we move the `createStakingPool` function to `CoverProducts.sol`
2. we remove the functionality to add products in the same tx (i.e. `stakingProducts().setInitialProducts(poolId, productInitParams);`) - the UX is similar - 2 txs instead of 1; we remove the dependency between SP and CP and we get rid of duplicate code in SP

Comment (danoctavian) [reply contracts/modules/cover/Cover.sol:580]: Removing dependency of SP on to CP is still there because it checks for requirePoolIsAllowed in setProducts and productsCount in recalculateEffectiveWeightsForAllProducts

Do you still see value in separating it into 2 transactions  and moving it to CoverProducts?

File: contracts/modules/legacy/LegacyPooledStaking.sol (hunk 2 above)
Comment (roxdanila) [reply contracts/modules/legacy/LegacyPooledStaking.sol:844]: agreed

File: contracts/modules/cover/Cover.sol (hunk 4 above)
Comment (shark0der) [reply contracts/modules/cover/Cover.sol:581]: There's one more issue with moving the `createStakingPool` function - currently the `Cover` is the operator that is allowed to call the `StakingPoolFactory` contract. The operator can be changed, however I don't think the overhead is worth it - `Cover` is also a source for some calls in the `StakingPool` contracts, size constraints are gone given the products being moved, do you also think it would be easier to just leave it here?

File: contracts/modules/cover/Cover.sol [hunk 5]
```diff
@@ -870,15 +661,16 @@ contract Cover is ICover, MasterAwareV2, IStakingPoolBeacon, ReentrancyGuard, Mu
...
 
-      if (productId >= _products.length) {
+      if (productId >= _coverProducts.productsCount()) {
         revert ProductDoesntExist();
       }
 
-      _initialPrices[i] = uint(_products[productId].initialPriceRatio);
-      _capacityReductionRatios[i] = uint(_products[productId].capacityReductionRatio);
+      _initialPrices[i] = uint(_coverProducts.products(productId).initialPriceRatio);
+      _capacityReductionRatios[i] = uint(_coverProducts.products(productId).capacityReductionRatio);
```
Comment (shark0der) [inline contracts/modules/cover/Cover.sol:673]: There are multiple calls to `coverProducts` here and almost nothing from the current contract. We need to move over this function as well. Additionally the product count should be cached into a local variable outside the for loop.

File: contracts/modules/cover/CoverProducts.sol [hunk 6]
```diff
@@ -0,0 +1,224 @@
...
+  mapping(uint => string) public productNames;
+  // productTypeId => productType name
+  mapping(uint => string) public productTypeNames;
+
+  // product id => allowed pool ids
+  mapping(uint => uint[]) public allowedPools;
+
+  /* ========== CONSTANTS ========== */
+
+  uint public constant GLOBAL_MIN_PRICE_RATIO = 100; // 1%
```
Comment (shark0der) [inline contracts/modules/cover/CoverProducts.sol:42]: This is also present in the Cover contract. Need to keep it in a single place. There is 1 usage here and 2 usages there, but likely to move one of the two here as well. Decide where to keep this constant and remove it on the other side.

File: contracts/modules/cover/CoverProducts.sol [hunk 7]
```diff
@@ -0,0 +1,224 @@
...
+
+      if (product.initialPriceRatio > PRICE_DENOMINATOR) {
+        revert InitialPriceRatioAbove100Percent();
+      }
+
+      if (product.capacityReductionRatio > CAPACITY_REDUCTION_DENOMINATOR) {
+        revert CapacityReductionRatioAbove100Percent();
+      }
+
+      // TODO: https://github.com/NexusMutual/smart-contracts/issues/859
```
Comment (shark0der) [inline contracts/modules/cover/CoverProducts.sol:116]: Probably a good moment to address this issue as well

File: contracts/modules/legacy/LegacyPooledStaking.sol [hunk 8]
```diff
@@ -840,160 +840,4 @@ contract LegacyPooledStaking is IPooledStaking, MasterAwareV2, PricesV1 {
...
-    );
-
-    uint deposit = stakers[stakerAddress].deposit;
-    require(deposit > 0, "Address has no stake to migrate");
-
-    INXMToken nxm = token();
-    uint nxmBalanceBefore = nxm.balanceOf(address(this));
-    uint expectedMigratedAmount = 0;
-
-    ProductInitializationParams[] memory products = getProducts(stakerAddress, deposit);
```
Comment (shark0der) [inline contracts/modules/legacy/LegacyPooledStaking.sol:907]: I think it's safe to remove the `getProducts` function as well?

File: contracts/modules/legacy/LegacyPooledStaking.sol [hunk 9]
```diff
@@ -840,160 +840,4 @@ contract LegacyPooledStaking is IPooledStaking, MasterAwareV2, PricesV1 {
...
-    require(actualMigratedAmount == expectedMigratedAmount, "NXM balance mismatch");
-
-    // Send unlocked NXM back
-    uint nxmToBeUnlocked = deposit - actualMigratedAmount;
-    if (nxmToBeUnlocked != 0) {
-      nxm.transfer(stakerAddress, nxmToBeUnlocked);
-    }
-  }
-
-  function migrateToExistingV2Pool(uint stakingPoolId, uint trancheId) external {
```
Comment (shark0der) [inline contracts/modules/legacy/LegacyPooledStaking.sol:992]: Please make sure the frontend is not using this function

File: contracts/modules/assessment/IndividualClaims.sol [hunk 10]
```diff
@@ -259,13 +264,13 @@ contract IndividualClaims is IIndividualClaims, MasterAwareV2 {
...
-    ICover coverContract = cover();
+    ICoverProducts coverProductsContract = coverProducts();
     CoverData memory coverData = cover().coverData(coverId);
     CoverSegment memory segment = cover().coverSegmentWithRemainingAmount(coverId, segmentId);
 
     {
-      Product memory product = coverContract.products(coverData.productId);
-      ProductType memory productType = coverContract.productTypes(product.productType);
+      Product memory product = coverProductsContract.products(coverData.productId);
+      ProductType memory productType = coverProductsContract.productTypes(product.productType);
```
Comment (shark0der) [inline contracts/modules/assessment/IndividualClaims.sol:273]: We seem to have a pattern of fetching the product and the type (saw this in some other contract). Consider creating a function `getProductWithType` to fetch both in a single call.

File: contracts/modules/staking/StakingPoolFactory.sol [hunk 11]
```diff
@@ -13,6 +13,8 @@ contract StakingPoolFactory is IStakingPoolFactory {
   // temporary beacon address storage to avoid constructor arguments in the proxy
   address public beacon;
 
+  address public coverAddress;
```
Comment (shark0der) [inline contracts/modules/staking/StakingPoolFactory.sol:16]: This contract is immutable and cannot be changed or replaced. Revert any changes here as they're neither possible nor required.

File: contracts/modules/staking/StakingProducts.sol (hunk 1 above)
Comment (shark0der) [reply contracts/modules/staking/StakingProducts.sol:618]: There can be only one operator so it cannot live in more than one contract. See other comment regarding the proposal to move it back to `Cover`.

File: contracts/modules/staking/StakingProducts.sol [hunk 12]
```diff
@@ -597,10 +607,87 @@ contract StakingProducts is IStakingProducts, MasterAwareV2, Multicall {
...
+    tokenController().assignStakingPoolManager(poolId, msg.sender);
+
+    _setInitialProducts(poolId, productInitParams);
+
+    return (poolId, stakingPoolAddress);
+  }
+
   /* dependencies */
 
+  function tokenController() internal view returns (ITokenController) {
```
Comment (shark0der) [inline contracts/modules/staking/StakingProducts.sol:675]: Dep only used in createStakingPool, remove after moving that one out.

File: contracts/modules/staking/StakingViewer.sol [hunk 13]
```diff
@@ -94,6 +95,10 @@ contract StakingViewer is Multicall {
   function stakingProducts() internal view returns (IStakingProducts) {
     return IStakingProducts(master.contractAddresses('SP'));
   }
+  
+  function coverProducts() internal view returns (ICoverProducts) {
+    return ICoverProducts(master.contractAddresses('CP'));
+  }
```
Comment (shark0der) [inline contracts/modules/staking/StakingViewer.sol:101]: This replaces `cover()` function entirely hence it can take its place

File: test/integration/utils/staking.js [hunk 14]
```diff
@@ -31,7 +31,7 @@ async function stake({ stakingPool, staker, productId, period, gracePeriod, amou
   // Stake to open up capacity
   await stakingPool.connect(staker).depositTo(
     stakingAmount,
-    firstTrancheId,
+    firstTrancheId + trancheIncrement,
```
Comment (shark0der) [inline test/integration/utils/staking.js:34]: `trancheIncrement` is never passed as an argument, no need to add it.

File: test/integration/setup.js [hunk 15]
```diff
@@ -169,9 +169,10 @@ async function setup() {
     coverNFTDescriptor.address,
   ]);
 
-  // 2. deploy Cover, StakingProducts and TokenController proxies
+  // 2. deploy Cover, StakingProducts, CoverProducts and TokenController proxies
   let cover = await deployProxy('Stub');
   let stakingProducts = await deployProxy('Stub');
+  let coverProducts = await deployProxy('Stub');
```
Comment (shark0der) [inline test/integration/setup.js:175]: We don't seem to need a stub for this contract, deploy it directly so we won't need an upgrade tx.

File: test/integration/setup.js [hunk 16]
```diff
@@ -275,8 +299,10 @@ async function setup() {
   };
 
   await cover.changeDependentContractAddress();
+  await stakingProducts.changeDependentContractAddress();
+  await coverProducts.changeDependentContractAddress();
```
Comment (shark0der) [inline test/integration/setup.js:303]: Is this actually needed?

File: contracts/mocks/Claims/CLMockCoverProducts.sol [hunk 17]
```diff
@@ -0,0 +1,40 @@
+// SPDX-License-Identifier: GPL-3.0-only
+
+pragma solidity ^0.8.18;
+
+import "../../interfaces/ICover.sol";
+import "../../interfaces/ICoverNFT.sol";
+
+
+contract CLMockCoverProducts {
```
Comment (shark0der) [inline contracts/mocks/Claims/CLMockCoverProducts.sol:9]: All mocks must inherit the interfaces now

File: contracts/mocks/Incidents/ICMockCoverProducts.sol [hunk 18]
```diff
@@ -0,0 +1,73 @@
+// SPDX-License-Identifier: GPL-3.0-only
+
+pragma solidity ^0.8.18;
+
+import "../../interfaces/ICover.sol";
+import "../../interfaces/ICoverNFT.sol";
+
+contract ICMockCoverProducts {
```
Comment (shark0der) [inline contracts/mocks/Incidents/ICMockCoverProducts.sol:8]: Please implement the interface as well

File: test/unit/IndividualClaims/setup.js [hunk 19]
```diff
@@ -48,6 +48,10 @@ async function setup() {
   const cover = await Cover.deploy(coverNFT.address);
   await cover.deployed();
 
+  const CoverProducts = await ethers.getContractFactory('ICMockCoverProducts');
+  const coverProducts = await CoverProducts.deploy();
+  await coverProducts.deployed();
+
   const Distributor = await ethers.getContractFactory('CLMockDistributor');
   const distributor = await Distributor.deploy(individualClaims.address);
   await distributor.deployed();
```
Comment (shark0der) [inline test/unit/IndividualClaims/setup.js:57]: Not in scope, but distributor seems to be unused. Let's nuke it?

**Review by shark0der — CHANGES_REQUESTED** [review]
See comments

File: contracts/modules/cover/Cover.sol (hunk 4 above)
Comment (danoctavian) [reply contracts/modules/cover/Cover.sol:581]: I'd prefer to move it out  to StakingProducts just to ensure we have plenty space in Cover.sol going forward and we don't have to think about its size again

Changing the tests to adapt to this involves lot of minute changes, so if we decide to move it out again we'd have to apply those changes again

File: test/integration/setup.js (hunk 16 above)
Comment (danoctavian) [reply test/integration/setup.js:303]: Yes if you remove those 2 lines - the tests start failing.

The storeddependencies are not there (0 address) so transactions fail.

File: contracts/modules/legacy/LegacyPooledStaking.sol (hunk 9 above)
Comment (danoctavian) [reply contracts/modules/legacy/LegacyPooledStaking.sol:992]: Checked - it's not used by frontend

File: contracts/modules/legacy/LegacyPooledStaking.sol (hunk 8 above)
Comment (danoctavian) [reply contracts/modules/legacy/LegacyPooledStaking.sol:907]: you're right, removed

File: contracts/modules/staking/StakingViewer.sol (hunk 13 above)
Comment (danoctavian) [reply contracts/modules/staking/StakingViewer.sol:97]: yes deleted

File: contracts/modules/staking/StakingProducts.sol (hunk 12 above)
Comment (danoctavian) [reply contracts/modules/staking/StakingProducts.sol:671]: pending decision to keep or move out

File: contracts/modules/staking/StakingProducts.sol [hunk 20]
```diff
@@ -597,10 +607,87 @@ contract StakingProducts is IStakingProducts, MasterAwareV2, Multicall {
     return surgePremium / allocationUnitsPerNxm;
   }
 
+  /* ========== STAKING POOL CREATION ========== */
+
+  function stakingPool(uint poolId) public view returns (IStakingPool) {
```
[the remark this answers is not in the document]
Comment (danoctavian) [reply contracts/modules/staking/StakingProducts.sol:608]: i kept `stakingPool` as the only function in the contract for convenience (it's the one used by all other contracts and tests)

File: contracts/modules/staking/StakingPool.sol [hunk 21]
```diff
@@ -180,7 +180,11 @@ contract StakingPool is IStakingPool, Multicall {
     uint _maxPoolFee,
     uint _poolId,
     string  calldata ipfsDescriptionHash
-  ) external onlyCoverContract {
+  ) external {
+
+    if (msg.sender != address(stakingProducts)) {
+      revert OnlyStakingProductsContract();
+    }
```
[the remark this answers is not in the document]
Comment (danoctavian) [reply contracts/modules/staking/StakingPool.sol:187]: pending createStakingPool movement decision

File: contracts/modules/assessment/IndividualClaims.sol (hunk 10 above)
Comment (danoctavian) [reply contracts/modules/assessment/IndividualClaims.sol]: applied

File: contracts/modules/cover/CoverProducts.sol (hunk 3 above)
Comment (danoctavian) [reply contracts/modules/cover/CoverProducts.sol:25]: added migration function

File: contracts/modules/staking/StakingPool.sol (hunk 21 above)
[the remark this answers is not in the document]
Comment (danoctavian) [reply contracts/modules/staking/StakingPool.sol:187]: Another note here:

All the changes here go out with the edit cover functionality.

That alters StakingPool regardless so we won't be eliminating an extra deploy here

File: contracts/modules/cover/Cover.sol [hunk 22]
```diff
@@ -35,9 +37,6 @@ contract Cover is ICover, MasterAwareV2, IStakingPoolBeacon, ReentrancyGuard, Mu
   // cover id => segment id => pool allocations array
   mapping(uint => mapping(uint => PoolAllocation[])) public coverSegmentAllocations;
 
-  // product id => allowed pool ids
-  mapping(uint => uint[]) public allowedPools;
```
Comment (shark0der) [inline contracts/modules/cover/Cover.sol:39]: breaks the storage layout

File: contracts/modules/staking/StakingPoolFactory.sol [hunk 23]
```diff
@@ -27,7 +27,7 @@ contract StakingPoolFactory is IStakingPoolFactory {
     return _stakingPoolCount;
   }
 
-  function create(address _beacon) external returns (uint poolId, address stakingPoolAddress) {
+  function create(address _beacon) public returns (uint poolId, address stakingPoolAddress) {
```
Comment (shark0der) [inline contracts/modules/staking/StakingPoolFactory.sol:30]: Revert this change as well please

File: test/integration/setup.js (hunk 16 above)
Comment (danoctavian) [reply test/integration/setup.js:301]: It's because contracts are added using the function initialize from our disposable:

https://github.com/NexusMutual/smart-contracts/blob/b4a1d0a51270eb817cfc67baa88810497cd9d4a5/contracts/mocks/Disposables/DisposableNXMaster.sol#L9


Not added through the standard function.

File: contracts/modules/cover/Cover.sol (hunk 22 above)
Comment (danoctavian) [reply contracts/modules/cover/Cover.sol:39]: addressed by adding those back

File: contracts/modules/staking/StakingPool.sol (hunk 21 above)
[the remark this answers is not in the document]
Comment (danoctavian) [reply contracts/modules/staking/StakingPool.sol:187]: We decided to keep createStakingPool for now in StakingProducts so we'll keep this change
